### PR TITLE
feat: support conv dynamo converter

### LINKED ISSUE #2250: Expose IConvolutionLayer in dynamo.conversion.impl

### PR DESCRIPTION
**Pull request by zewenli98**
# Description

Support conv dynamo converter.

Fixes #2250, #2185

## Type of change

- New feature (non-breaking change which adds functionality)

# Checklist:

- [x] My code follows the style guidelines of this project (You can use the linters)
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas and hacks
- [ ] I have made corresponding changes to the documentation
- [x] I have added tests to verify my fix or my feature
- [x] New and existing unit tests pass locally with my changes
- [x] I have added the relevant labels to my PR in so that relevant reviewers are notified

### CODE REVIEW

File: py/torch_tensorrt/dynamo/conversion/impl/conv.py [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+
+    # If the bias is a TRTTensor, set it as an input of the layer
+    if isinstance(bias, TRTTensor):
+        conv_layer.set_input(2, bias)
+
+    # Expand parameters manually for Conv1D computations
+    if is_conv1d:
+        padding = tuple(padding) + (0,)
+        stride = extend_attr_to_tuple(stride, 2)
+        dilation = extend_attr_to_tuple(dilation, 2)
```
Comment (gs-olive) [inline py/torch_tensorrt/dynamo/conversion/impl/conv.py:104]: Could you also address the issue in #2185 in this PR?

File: py/torch_tensorrt/dynamo/conversion/impl/conv.py [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+        assert input.shape[1] != -1, "Channel dim can't be dynamic for convolution."
+
+    if is_conv1d:
+        # Apply an unsqueeze operation to transform the conv1d problem into conv2d
+        input = aten_ops_converters.aten_ops_unsqueeze(
+            network, target, (input, -1), {}, name + "_unsqueeze"
+        )
+
+    # Process bias terms
+    if isinstance(bias, torch.Tensor):
```
Comment (gs-olive) [inline py/torch_tensorrt/dynamo/conversion/impl/conv.py:49]: Switch to `isinstance(bias, (torch.Tensor, np.ndarray))`

File: py/torch_tensorrt/dynamo/conversion/impl/conv.py [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+    # Process weight terms
+    if network.has_explicit_precision or isinstance(weight, TRTTensor):
+        weight = get_trt_tensor(network, weight, f"{name}_weight")
+        # Append new dimension (unsqueeze) if the convolution is 1d
+        if is_conv1d:
+            weight = aten_ops_converters.aten_ops_unsqueeze(
+                network, target, (weight, -1), {}, name + "_unsqueeze_weight"
+            )
+
+    elif isinstance(weight, torch.Tensor):
```
Comment (gs-olive) [inline py/torch_tensorrt/dynamo/conversion/impl/conv.py:70]: Switch to `isinstance(weight, (torch.Tensor, np.ndarray))`

File: py/torch_tensorrt/dynamo/conversion/impl/conv.py (hunk 1 above)
Comment (gs-olive) [reply py/torch_tensorrt/dynamo/conversion/impl/conv.py:104]: Specifically, you could duplicate and fix this function from the FX converter utilities into `dynamo/conversion/converter_utils.py`:
https://github.com/pytorch/TensorRT/blob/b3089bf00ec2e5124138201f6d76c8e3b6636b64/py/torch_tensorrt/fx/converters/converter_utils.py#L128-L146

**Review by gs-olive — COMMENTED** [review]
Great to have this converter in Dynamo. Made some suggestions for fixes from issues encountered in the FX path.

File: py/torch_tensorrt/dynamo/conversion/impl/conv.py (hunk 1 above)
Comment (zewenli98) [reply py/torch_tensorrt/dynamo/conversion/impl/conv.py:104]: Done. Thanks for the suggestion!

File: py/torch_tensorrt/fx/converters/converter_utils.py
Comment (gs-olive) [inline py/torch_tensorrt/fx/converters/converter_utils.py:1]: No need to make any changes here to the `fx/` directory since the function has been duplicated into the `dynamo/` directory.

**Review by gs-olive — COMMENTED** [review]
Added a small comment on the changes to the FX directory. After that, just needs a rebase and should be good to go!

**Comment by zewenli98** [comment]
Thanks! rebased!

File: py/torch_tensorrt/dynamo/conversion/aten_ops_converters.py [hunk 4]
```diff
@@ -812,3 +812,34 @@ def aten_ops_isinf(
...
+    args: Tuple[Argument, ...],
+    kwargs: Dict[str, Argument],
+    name: str,
+) -> Union[TRTTensor, Sequence[TRTTensor]]:
+    # we do not handle transposed.
+    if args[6] is True:
+        raise RuntimeError(f"Target {target} does not support `transposed=True` ")
+    # we do not handle output_padding.
+    if args[7] not in ([0], [0, 0], [0, 0, 0]):
+        raise RuntimeError(f"Target {target} has non-0 output_padding")
```
Comment (gs-olive) [inline py/torch_tensorrt/dynamo/conversion/aten_ops_converters.py:830]: Could these be raised into a capability validator, so they don't throw errors? For instance, you could add:
```python
def convolution_check(convolution_node: Node) -> bool:
    return (not args[6]) and (args[7] in ([0], [0, 0], [0, 0, 0]))

@dynamo_tensorrt_converter(torch.ops.aten.convolution.default, capability_validator=convolution_check)
```

File: py/torch_tensorrt/dynamo/conversion/impl/conv.py [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+        # Assume the dtype of activation is torch.quint8
+        mark_as_int8_layer(conv_layer, get_dyn_range(scale, zero_point, torch.quint8))
+
+    result = conv_layer.get_output(0)
+
+    if is_conv1d:
+        # Apply a squeeze operation to transform the conv2d problem back into conv1d
+        result = aten_ops_converters.aten_ops_squeeze(
+            network, target, (result, -1), {}, name + "_squeeze"
+        )
```
Comment (gs-olive) [inline py/torch_tensorrt/dynamo/conversion/impl/conv.py:127]: Change this to invoke the `squeeze` implementation directly, instead of the `squeeze` converter. For instance:
```python
impl.squeeze.squeeze(network, target, source_ir, name + "_squeeze_conv1d", result, -1)
```

File: py/torch_tensorrt/dynamo/conversion/impl/conv.py [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+
+    # Process weight terms
+    if network.has_explicit_precision or isinstance(weight, TRTTensor):
+        weight = get_trt_tensor(network, weight, f"{name}_weight")
+        # Append new dimension (unsqueeze) if the convolution is 1d
+        if is_conv1d:
+            weight = aten_ops_converters.aten_ops_unsqueeze(
+                network, target, (weight, -1), {}, name + "_unsqueeze_weight"
+            )
```
Comment (gs-olive) [inline py/torch_tensorrt/dynamo/conversion/impl/conv.py:68]: Change this to invoke the `unsqueeze` implementation directly, instead of the `unsqueeze` converter. Generally, converters should not be called directly from code (only their underlying base implementation should). For instance:
```python
impl.unsqueeze.unsqueeze(network, target, source_ir, name + "_unsqueeze_weight", weight, -1)
```

File: py/torch_tensorrt/dynamo/conversion/impl/conv.py [hunk 7]
```diff
@@ -0,0 +1,129 @@
+from typing import Optional, Sequence, Union
+
+import numpy as np
+
+# @manual=//deeplearning/trt/python:py_tensorrt
+import tensorrt as trt
+import torch
+from torch.fx.node import Target
+from torch_tensorrt.dynamo.conversion import aten_ops_converters
```
Comment (gs-olive) [inline py/torch_tensorrt/dynamo/conversion/impl/conv.py:9]: Switch to `from torch_tensorrt.dynamo.conversion import impl` (see below)

**Review by gs-olive — COMMENTED** [review]
See below comments on calling convention for converters, then should be good to go!

File: py/torch_tensorrt/dynamo/conversion/impl/conv.py (hunk 7 above)
Comment (zewenli98) [reply py/torch_tensorrt/dynamo/conversion/impl/conv.py]: @gs-olive Thank you so much for the suggestions and detailed explanations! These issues are fixed!

**Review by gs-olive — APPROVED** [review]
Looks great! Approved, pending CI
